src/hardware/vga_s3.cpp

```diff
@@ -822,4 +829,106 @@ void SVGA_Setup_S3Trio(void)
+	{
+		return regnum;
+	}
+
+	bool OverrideReadRegister(uint8_t regnum, uint8_t* rval, uint8_t* rval_mask)
+	{
+		return false;
+	}
+
+	Bits ParseWriteRegister(uint8_t regnum, uint8_t value)
+	{
+		if ((regnum>=0x18) && (regnum<0x28)) return -1;	// base addresses are read-only
+		if ((regnum>=0x30) && (regnum<0x34)) return -1;	// expansion rom addresses are read-only
+		switch (regnum) {
+			case 0x10:
+				return (PCI_GetCFGData(PCIId(), PCISubfunction(), 0x10)&0x0f);
+			case 0x11:
+				return 0x00;
+			case 0x12:
+				//return (value&0xc0);	// -> 4mb addressable
+				return (value&0x00);	// -> 16mb addressable
+			case 0x13:
+				return value;
+			case 0x14:
+				return (PCI_GetCFGData(PCIId(), PCISubfunction(), 0x10)&0x0f);
+			case 0x15:
+				return 0x00;
+			case 0x16:
+				return value;	// -> 64kb addressable
+			case 0x17:
+				return value;
+			default:
+				break;
+		}
+		return value;
+	}
+
+	bool InitializeRegisters(uint8_t registers[256])
+	{
+		// init (S3 graphics card)
+		//registers[0x08] = 0x44;	// revision ID (s3 trio64v+)
+		registers[0x08] = 0x00;	// revision ID
+		registers[0x09] = 0x00;	// interface
+		registers[0x0a] = 0x00;	// subclass type (vga compatible)
+		//registers[0x0a] = 0x01;	// subclass type (xga device)
+		registers[0x0b] = 0x03;	// class type (display controller)
+		registers[0x0c] = 0x00;	// cache line size
+		registers[0x0d] = 0x00;	// latency timer
+		registers[0x0e] = 0x00;	// header type (other)
+
+		// reset
+		registers[0x04] = 0x23;	// command register (vga palette snoop, ports enabled, memory space enabled)
+		registers[0x05] = 0x00;
+		registers[0x06] = 0x80;	// status register (medium timing, fast back-to-back)
+		registers[0x07] = 0x02;
+
+		//registers[0x3c] = 0x0b;	// irq line
+		//registers[0x3d] = 0x01;	// irq pin
+
+		// uint32_t gfx_address_space=(((uint32_t)S3_LFB_BASE)&0xfffffff0)
+		// | 0x08;	// memory space, within first 4GB, prefetchable
+		uint32_t gfx_address_space = (((uint32_t)S3_LFB_BASE) &
+		                              0xfffffff0); // memory space,
+		                                           // within first 4GB
+		registers[0x10] = (uint8_t)(gfx_address_space & 0xff); // base
+		                                                       // addres 0
+		registers[0x11] = (uint8_t)((gfx_address_space >> 8) & 0xff);
+		registers[0x12] = (uint8_t)((gfx_address_space >> 16) & 0xff);
+		registers[0x13] = (uint8_t)((gfx_address_space >> 24) & 0xff);
+
+		uint32_t gfx_address_space_mmio = (((uint32_t)S3_LFB_BASE + 0x1000000) &
+		                                   0xfffffff0); // memory space,
+		                                                // within first 4GB
+		registers[0x14] = (uint8_t)(gfx_address_space_mmio & 0xff); // base addres 0
+		registers[0x15] = (uint8_t)((gfx_address_space_mmio >> 8) & 0xff);
+		registers[0x16] = (uint8_t)((gfx_address_space_mmio >> 16) & 0xff);
+		registers[0x17] = (uint8_t)((gfx_address_space_mmio >> 24) & 0xff);
```

Comment:
I'm not a fan of this "direct access" thing... I'd introduce `write_register` functions or something.

Reply:
We have a bunch of functions like that in headers `mem_host.h` and `mem_unaligned.h` (those are our cleanups though and were not available in vanilla dosbox). I hope one of them will be ok in this context. I will probably push it the future PRs though.